### ISSUE TITLE
React 18 implicit children typing issue?

### ISSUE THREAD

**Issue by girvo** [mantinedev/mantine#1245]
### What package has an issue

@mantine/core

### Describe the bug

So the `@types/react` typing package for TypeScript, for React 18, has remove the implicit children for all component types, and this is causing me to run into some issues with our extensions of some Mantine components.

```typescript
import React, { ComponentType, Key, Fragment } from 'react'

// given a Props interface with this:
  header?: () => React.ReactNode
}

// snip
  return (
    <div className={classes.grid}>
      {props.header && (
        <MediaQuery largerThan="sm" styles={hide}>
          {props.header()}
        </MediaQuery>
      )}
```

When used with React 18.0.1, `@types/react` 18.0.1 and `@mantine/core` v4.1.3, this gives an error about React.ReactNode not being React.ReactNode!

At the `{props.header()}` line:
```
Type 'React.ReactNode' is not assignable to type 'import("/workspaces/cobalt-dev/node_modules/@types/react/index").ReactNode'.
  Type '{}' is not assignable to type 'ReactNode'.
  ```

I'm not entirely sure what is causing this, other than some kind of typing mismatch when trying to use React v18 and it's v18 @types package with Mantine?

### In which browser did the problem occur

All browsers

**Comment by fl0wx** [comment]
@girvo yesterday i found this, which could be related: https://www.reddit.com/r/reactjs/comments/u2rbbx/library_authors_should_revise_typings_to_support/

**Comment by rtivital** [comment]
Mantine does not use FC type anywhere, so most likely provided codemod will not help.

**Comment by girvo** [comment]
@rtivital Yeah I initially assumed it was something to do with that as well initially (you can tell in my initial paragraph thats what I thought it was) but then checked the source and saw that the FC type isn't used anywhere, and yet the type error still exists. Super odd.

Is there anything I can do to help with tracking this down? It's unfortunately prevented us from migrating to v18 at work. Happy to do some more debugging or see if I can reproduce a tiny test case for it if need be, or post more of the source we have for it, but it's super weird. Basically thinks it's `React.ReactNode` type is the empty object, instead of the actual `ReactNode` definition, despite both my code and Mantine ostensibly using the same `@types/react` package

**Comment by wes337** [comment]
> Mantine does not use FC type anywhere, so most likely provided codemod will not help.

Are you sure? I ran the codemod (https://github.com/eps1lon/types-react-codemod) and quite a few changes were made:

![image](https://user-images.githubusercontent.com/9464690/163926225-acda8eb0-13d9-4480-8539-521c83cc0555.png)

**Comment by wes337** [comment]
@girvo Have you tried adding React 17 to the resolutions of your package.json?

```json
"resolutions": {
    "@types/react": "17.0.8"
}
```

**Comment by girvo** [comment]
@wes337 While using `react@18.0.1` as-the-library itself?

**Comment by wes337** [comment]
I think so. I was just reading in that issue you linked, and many people found a solution using resolutions: https://github.com/facebook/react/issues/24304#issuecomment-1093458894.

**Comment by spigelli** [comment]
Also experiencing

> TS2322: Type '{ children: Element[]; size: "sm"; style: { height: string; }; }' is not assignable to type 'IntrinsicAttributes & ContainerProps & RefAttributes<HTMLDivElement>'.   Property 'children' does not exist on type 'IntrinsicAttributes & ContainerProps & RefAttributes<HTMLDivElement>'.


With versions:
```json
    "@types/react": "18.0.13",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "@mantine/core": "^4.2.11",
```

**Comment by rtivital** [comment]
Fixed in 5.0
